pilot/pkg/serviceregistry/kube/controller/network.go

```diff
@@ -313,6 +333,86 @@ func (n *networkManager) getGatewayDetails(svc *model.Service) []model.NetworkGa
 	return nil
 }
 
+func (n *networkManager) watchGatewayResources(c *Controller, stop <-chan struct{}) {
+	n.gatewayResourceClient = kclient.NewDelayedInformer(c.client, gvr.KubernetesGateway, kubetypes.StandardInformer, kubetypes.Filter{})
+	registerHandlers(c, n.gatewayResourceClient, "Gateways", n.handleGatewayResource, nil)
+	n.gatewayResourceClient.Start(stop)
+}
+
+// handleGateway resource adds a NetworkGateway for each combination of address and auto-passthrough listener
+// discovering duplicates from the generated Service is not a huge concern as we de-duplicate in NetworkGateways
+// which returns a set, although it's not totally efficient.
+func (n *networkManager) handleGatewayResource(_ controllers.Object, obj controllers.Object, event model.Event) error {
+	gw, ok := obj.(*v1beta1.Gateway)
+	if !ok {
+		return nil
+	}
+
+	if nw := gw.GetLabels()[label.TopologyNetwork.Name]; nw == "" {
+		return nil
+	}
+
+	// we don't want to discover gateways with class "istio-remote" from outside cluster's API servers.
+	// this will only pass for the "config cluster" or "local cluster"
+	if !n.discoverRemoteGatewayResources && gw.Spec.GatewayClassName == "istio-remote" {
+		return nil
+	}
+
+	gatewaysChanged := false
+	n.Lock()
+	defer func() {
+		n.Unlock()
+		if gatewaysChanged {
+			n.NotifyGatewayHandlers()
+		}
+	}()
+
+	previousGateways := n.gatewaysFromResource[gw.UID]
+
+	if event == model.EventDelete {
+		gatewaysChanged = len(previousGateways) > 0
+		delete(n.gatewaysFromResource, gw.UID)
+		return nil
+	}
+
+	autoPassthrough := func(l v1beta1.Listener) bool {
+		return kube.IsAutoPassthrough(gw.GetLabels(), l)
+	}
+
+	base := model.NetworkGateway{
+		Network: network.ID(gw.GetLabels()[label.TopologyNetwork.Name]),
+		Cluster: n.clusterID,
+	}
+	newGateways := model.NetworkGatewaySet{}
+	for _, addr := range gw.Spec.Addresses {
+		for _, l := range slices.Filter(gw.Spec.Listeners, autoPassthrough) {
+			networkGateway := base
+			networkGateway.Addr = addr.Value
+			networkGateway.Port = uint32(l.Port)
```

Comment:
for what case can it have multiple ports for network gateway

Reply:
understand, I am ok with supporting multi ports, in future we could use different port proxying TCP protocol

Reply:
Eventually we will probably start to support different types of port like HBONE